File: libs/java/dal/db/dal-db/src/main/java/ca/bc/gov/tno/dal/db/entities/Schedule.java
package ca.bc.gov.tno.dal.db.entities;

import java.util.ArrayList;
import java.util.Date;
import java.util.EnumSet;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.Convert;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.OneToMany;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;

import com.fasterxml.jackson.annotation.JsonBackReference;

import ca.bc.gov.tno.dal.db.AuditColumns;
import ca.bc.gov.tno.dal.db.Months;
import ca.bc.gov.tno.dal.db.ScheduleType;
import ca.bc.gov.tno.dal.db.WeekDays;
import ca.bc.gov.tno.dal.db.converters.WeekDaysAttributeConverter;
import ca.bc.gov.tno.dal.db.converters.MonthsAttributeConverter;

/**
 * Schedule class, provides a way to manage scheduling details for data sources.
 */
@Entity
@Table(name = "schedule", schema = "public")
public class Schedule extends AuditColumns {
  /**
   * Primary key to identify the schedule.
   */
  @Id
  @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "seq_schedule")
  @SequenceGenerator(name = "seq_schedule", allocationSize = 1)
  @Column(name = "id", nullable = false)
  private int id;

  /**
   * A unique name to identify the schedule.
   */
  @Column(name = "name", nullable = false)
  private String name;

  /**
   * A description of the schedule.
   */
  @Column(name = "description")
  private String description = "";

  /**
   * Whether this record is enabled or disabled.
   */
  @Column(name = "is_enabled", nullable = false)
  private boolean enabled = true;

  /**
   * The type of schedule.
   */
  @Column(name = "schedule_type", nullable = false)
  private ScheduleType scheduleType = ScheduleType.Repeating;

  /**
   * The number of milliseconds the service should rest before running again.
   */
  @Column(name = "delay_ms", nullable = false)
  private int delayMS;

  /**
   * The date and time the service should begin running on. This is useful if a
   * service should be delayed from running for a period of time.
   * null = Run immediately.
   */
  @Temporal(TemporalType.TIMESTAMP)
  @Column(name = "run_on")
  private Date runOn;

  /**
   * At what time the schedule should start.
   */
  @Temporal(TemporalType.TIME)
  @Column(name = "start_at")
  private Date startAt;

  /**
   * At what time the schedule should stop.
   */
  @Temporal(TemporalType.TIME)
  @Column(name = "stop_at")
  private Date stopAt;

  /**
   * Number of times to run before waiting for next RunAt.
   * 0 = is used for continuous running.
   */
  @Column(name = "repeat", nullable = false)
  private int repeat;

  /**
   * Identify which week days the service should run.
   * NA = Do not use runOnWeekDays.
   */
  @Column(name = "run_on_week_days", nullable = false)
  @Convert(converter = WeekDaysAttributeConverter.class)
  private EnumSet<WeekDays> runOnWeekDays;

  /**
   * Identify which months the service should run.
   * NA = Do not use runOnMonths.
   */
  @Column(name = "run_on_months", nullable = false)
  @Convert(converter = MonthsAttributeConverter.class)
  private EnumSet<Months> runOnMonths;

  /**
   * Identify the day of the month the service should run.
   * 0 = Do not use dayOfMonth.
   */
  @Column(name = "day_of_month", nullable = false)
  private int dayOfMonth;

  /**
   * A collection of data sources that belong to this schedule.
   */
  @JsonBackReference("data_sources")
  @OneToMany(mappedBy = "schedule", fetch = FetchType.LAZY)
  private List<DataSourceSchedule> dataSourceSchedules = new ArrayList<>();

  /**
   * Creates a new instance of a Schedule object.
   */
  public Schedule() {

  }

  /**
   * Creates a new instance of a Schedule object, initializes with specified
   * parameters.
   *
   * @param id   Primary key
   * @param name Unique name
   */
  public Schedule(int id, String name) {
    this.id = id;
    this.name = name;
  }

  /**
   * @return int return the id
   */
  public int getId() {
    return id;
  }

  /**
   * @param id the id to set
   */
  public void setId(int id) {
    this.id = id;
  }

  /**
   * @return String return the name
   */
  public String getName() {
    return name;
  }

  /**
   * @param name the name to set
   */
  public void setName(String name) {
    this.name = name;
  }

  /**
   * @return String return the description
   */
  public String getDescription() {
    return description;
  }

  /**
   * @param description the description to set
   */
  public void setDescription(String description) {
    this.description = description;
  }

  /**
   * @return boolean return the enabled
   */
  public boolean isEnabled() {
    return enabled;
  }

  /**
   * @param enabled the enabled to set
   */
  public void setEnabled(boolean enabled) {
    this.enabled = enabled;
  }

  /**
   * @return int return the delayMS
   */
  public int getDelayMS() {
    return delayMS;
  }

  /**
   * @param delayMS the delayMS to set
   */
  public void setDelayMS(int delayMS) {
    this.delayMS = delayMS;
  }

  /**
   * @return Date return the runOn
   */
  public Date getRunOn() {
    return runOn;
  }

  /**
   * @param runOn the runOn to set
   */
  public void setRunOn(Date runOn) {
    this.runOn = runOn;
  }

  /**
   * @return Date return the repeat
   */
  public int getRepeat() {
    return repeat;
  }

  /**
   * @param repeat the repeat to set
   */
  public void setRepeat(int repeat) {
    this.repeat = repeat;
  }

  /**
   * @return EnumSet{WeekDays} return the runOnWeekDays
   */
  public EnumSet<WeekDays> getRunOnWeekDays() {
    return runOnWeekDays;
  }

  /**
   * @param runOnWeekDays the runOnWeekDays to set
   */
  public void setRunOnWeekDays(EnumSet<WeekDays> runOnWeekDays) {
    this.runOnWeekDays = runOnWeekDays;
  }

  /**
   * @return EnumSet{Months} return the runOnMonths
   */
  public EnumSet<Months> getRunOnMonths() {
    return runOnMonths;
  }

  /**
   * @param runOnMonths the runOnMonths to set
   */
  public void setRunOnMonths(EnumSet<Months> runOnMonths) {
    this.runOnMonths = runOnMonths;
  }

  /**
   * @return int return the dayOfMonth
   */
  public int getDayOfMonth() {
    return dayOfMonth;
  }

  /**
   * @param dayOfMonth the dayOfMonth to set
   */
  public void setDayOfMonth(int dayOfMonth) {
    this.dayOfMonth = dayOfMonth;
  }

  /**
   * @return ScheduleType return the scheduleType
   */
  public ScheduleType getScheduleType() {
    return scheduleType;
  }

  /**
   * @param scheduleType the scheduleType to set
   */
  public void setScheduleType(ScheduleType scheduleType) {
    this.scheduleType = scheduleType;
  }

  /**
   * @return Date return the startAt
   */
  public Date getStartAt() {
    return startAt;
  }

  /**
   * @param startAt the startAt to set
   */
  public void setStartAt(Date startAt) {
    this.startAt = startAt;
  }

  /**
   * @return Date return the stopAt
   */
  public Date getStopAt() {
    return stopAt;
  }

  /**
   * @param stopAt the stopAt to set
   */
  public void setStopAt(Date stopAt) {
    this.stopAt = stopAt;
  }

  /**
   * @return List{DataSourceSchedule} return the dataSourceSchedules
   */
  public List<DataSourceSchedule> getDataSourceSchedules() {
    return dataSourceSchedules;
  }

  /**
   * @param dataSourceSchedules the dataSourceSchedules to set
   */
  public void setDataSourceSchedules(List<DataSourceSchedule> dataSourceSchedules) {
    this.dataSourceSchedules = dataSourceSchedules;
  }

}

File: libs/net/entities/Schedule.cs
using System.ComponentModel.DataAnnotations.Schema;

namespace TNO.Entities;

/// <summary>
/// Schedule class, provides an entity model for a schedule configuration in the database.
/// This provides a way to control when and how often an action (i.e. ingest) will run.
/// </summary>
[Table("schedule")]
public class Schedule : AuditColumns
{
    #region Properties
    /// <summary>
    /// get/set - The primary key.
    /// </summary>
    [Column("id")]
    public int Id { get; set; }

    /// <summary>
    /// get/set - The name of the schedule.
    /// </summary>
    [Column("name")]
    public string Name { get; set; } = "";

    /// <summary>
    /// get/set - The schedule description.
    /// </summary>
    [Column("description")]
    public string Description { get; set; } = "";

    /// <summary>
    /// get/set - Whether the schedule is enabled.
    /// </summary>
    [Column("is_enabled")]
    public bool IsEnabled { get; set; } = true;

    /// <summary>
    /// get/set - the type of schedule.
    /// </summary>
    [Column("schedule_type")]
    public ScheduleType ScheduleType { get; set; } = ScheduleType.Continuous;

    /// <summary>
    /// get/set - Number of milliseconds to wait between each execution of the action.
    /// </summary>
    [Column("delay_ms")]
    public int DelayMS { get; set; }

    /// <summary>
    /// get/set - The date and time to start executing the action (based on StartAt).
    /// </summary>
    [Column("run_on")]
    public DateTime? RunOn { get; set; }

    /// <summary>
    /// get/set - The time to start the action.
    /// </summary>
    [Column("start_at")]
    public TimeSpan? StartAt { get; set; }

    /// <summary>
    /// get/set - The time to stop the action.
    /// </summary>
    [Column("stop_at")]
    public TimeSpan? StopAt { get; set; }

    /// <summary>
    /// get/set - Whether this schedule will be deleted after a successful execution.
    /// </summary>
    [Column("run_only_once")]
    public bool RunOnlyOnce { get; set; }

    /// <summary>
    /// get/set - The number of times to execute the action.  This is only relevant for continuous schedules.
    /// </summary>
    [Column("repeat")]
    public int Repeat { get; set; }

    /// <summary>
    /// get/set - The day of the week to run on.
    /// </summary>
    [Column("run_on_week_days")]
    public ScheduleWeekDay RunOnWeekDays { get; set; } = ScheduleWeekDay.NA;

    /// <summary>
    /// get/set - The month to run on.
    /// </summary>
    [Column("run_on_months")]
    public ScheduleMonth RunOnMonths { get; set; } = ScheduleMonth.NA;

    /// <summary>
    /// get/set - The day of the month to run on.
    /// </summary>
    [Column("day_of_month")]
    public int DayOfMonth { get; set; }

    /// <summary>
    /// get/set - Foreign key to user who requested this content to be generated.
    /// </summary>
    [Column("requested_by_id")]
    public int? RequestedById { get; set; }

    /// <summary>
    /// get/set - The user who requested this content to be generated.
    /// </summary>
    public User? RequestedBy { get; set; }

    /// <summary>
    /// get - Collection of ingests this schedule belongs to.
    /// </summary>
    public virtual List<Ingest> Ingests { get; set; } = new List<Ingest>();

    /// <summary>
    /// get/set - Collection of ingests this schedule belongs to (many-to-many).
    /// </summary>
    public virtual List<IngestSchedule> IngestsManyToMany { get; set; } = new List<IngestSchedule>();
    #endregion

    #region Constructors
    /// <summary>
    /// Creates a new instance of a Schedule object.
    /// </summary>
    protected Schedule() { }

    /// <summary>
    /// Creates a new instance of a Schedule object, initializes with specified parameters.
    /// </summary>
    /// <param name="name"></param>
    /// <param name="scheduleType"></param>
    /// <param name="delayMS"></param>
    /// <exception cref="ArgumentException"></exception>
    public Schedule(string name, ScheduleType scheduleType, int delayMS)
    {
        if (String.IsNullOrWhiteSpace(name)) throw new ArgumentException("Parameter is required and cannot be null, empty, or whitespace", nameof(name));

        this.Name = name;
        this.ScheduleType = scheduleType;
        this.DelayMS = delayMS;
    }
    #endregion
}
